pkg/sget/blob.go

```diff
@@ -0,0 +1,214 @@
+	signatureoptions "github.com/sigstore/sigstore/pkg/signature/options"
+)
+
+func (sg *SecureGet) GetBlob(ctx context.Context, sigRef, artifactRef string) error {
+	var pubKey sigstoresigs.Verifier
+	var err error
+	var cert *x509.Certificate
+
+	artifact, err := blob.LoadFileOrURL(artifactRef)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+		artifact = []byte(artifactRef)
+	}
+
+	if sigRef == "" {
+		sigRef = artifactRef + ".sig"
+	}
+
+	targetSig, err := blob.LoadFileOrURL(sigRef)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			// ignore if file does not exist, it can be a base64 encoded string as well
+			return err
+		}
+		targetSig = []byte(sigRef)
+	}
+
+	if isb64(targetSig) {
+		targetSig, err = base64.StdEncoding.DecodeString(string(targetSig))
+		if err != nil {
+			return err
+		}
+	}
+
+	rClient, err := rekor.NewClient(sg.RekorURL)
+	if err != nil {
+		return err
+	}
+
+	uuids, err := cosign.FindTLogEntriesByPayload(ctx, rClient, artifact)
+	if err != nil {
+		return err
+	}
+
+	if len(uuids) == 0 {
+		return errors.New("could not find a tlog entry for provided blob")
+	}
+
+	tlogEntry, err := cosign.GetTlogEntry(ctx, rClient, uuids[0])
+	if err != nil {
+		return err
+	}
+
+	certs, err := extractCerts(tlogEntry)
+	if err != nil {
+		return err
+	}
+	cert = certs[0]
+	pubKey, err = sigstoresigs.LoadECDSAVerifier(cert.PublicKey.(*ecdsa.PublicKey), crypto.SHA256)
+	if err != nil {
+		return err
+	}
+
+	if err := pubKey.VerifySignature(bytes.NewReader(targetSig), bytes.NewReader(artifact)); err != nil {
+		return err
+	}
+
+	if cert != nil { // cert
+		if err := cosign.TrustedCert(cert, fulcio.GetRoots()); err != nil {
+			return err
+		}
+		fmt.Fprintln(os.Stderr, "Certificate is trusted by Fulcio Root CA")
+		fmt.Fprintln(os.Stderr, "Email:", cert.EmailAddresses)
+		for _, uri := range cert.URIs {
+			fmt.Fprintf(os.Stderr, "URI: %s://%s%s\n", uri.Scheme, uri.Host, uri.Path)
+		}
+		fmt.Fprintln(os.Stderr, "Issuer: ", sigs.CertIssuerExtension(cert))
+	}
+	fmt.Fprintln(os.Stderr, "Verified OK")
+
+	rekorClient, err := rekor.NewClient(sg.RekorURL)
+	if err != nil {
+		return err
+	}
+	var pubBytes []byte
+
+	pubBytes, err = sigs.PublicKeyPem(pubKey, signatureoptions.WithContext(ctx))
+	if err != nil {
+		return err
+	}
+
+	if cert != nil {
+		pubBytes, err = cryptoutils.MarshalCertificateToPEM(cert)
+		if err != nil {
+			return err
+		}
+	}
+
+	b64sig := base64.StdEncoding.EncodeToString(targetSig)
+	if err != nil {
+		return err
+	}
+
+	uuid, index, err := cosign.FindTlogEntry(ctx, rekorClient, b64sig, artifact, pubBytes)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(os.Stderr, "tlog entry verified with uuid: %q index: %d\n", uuid, index)
+
+	_, err = sg.Out.Write(artifact)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// TODO(shibumi): we might want to export this function or add this to our package?!
+func extractCerts(e *models.LogEntryAnon) ([]*x509.Certificate, error) {
```

Comment:
Does it make sense to export this function and add it to another package and then re-import the function here?
Right now we are having the code twice (once in the sget code and once in the cosign code).